tests/test_view.py

```diff
@@ -59,4 +60,156 @@ def test_pycon(self):
         self.assertFalse(view.mgr.pycon.python_redirect)
 
 
+def angle_axis(angle_deg, axis):
+    a = axis
+    angle = np.radians(angle_deg)
+
+    cos_angle = np.cos(angle)
+    sin_angle = np.sin(angle)
+
+    rotation = np.identity(3) * cos_angle
+
+    m = np.outer(a, a) * (1 - cos_angle)
+    n = np.array([
+        [0, -a[2], a[1]],
+        [a[2], 0, -a[0]],
+        [-a[1], a[0], 0]
+    ]) * sin_angle
+
+    rotation += m + n
+
+    return rotation
+
+
+@unittest.skip("GUI is not yet available for testing")
+class ViewFPSCameraTC(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        widget = view.RManager.instance.setUp().add3DWidget()
+        widget.setCameraType("fps")
+
+        cls.widget = widget
+        cls.controller = widget.cameraController()
+
+        cls.move = cls.controller.updateCameraPosition
+        cls.get_pos = cls.controller.position
+
+    def test_translation(self):
+        dt = 0.01
+        linear_speed = self.controller.linear_speed()
+
+        delta = dt * linear_speed
+        new_pos = np.array(self.get_pos()) + [delta, delta, -delta]
+
+        self.move(x=1, dt=dt)
+        self.assertAlmostEqual(self.get_pos()[0], new_pos[0], delta=1e-2)
```

Comment:
Shouldn't the position be exact?

Reply:
I wish it was possible, but there's always a little error in calculations, because qt takes cross products and magnitude of vectors in [qcamera.translate](https://github.com/qt/qt3d/blob/82b59c89e282dcc04a01727e355d284c128752ab/src/render/frontend/qcamera.cpp). But we don't need very precise value here, delta=1e-2 is pretty enough

Reply:
A smaller tolerance close to the round-off done by the helper should be used.  1 percentage does not look like a small tolerance.  If possible, please use the smallest possible tolerance.  Alternately, please elaborate why 1 percent is sufficiently smaller.

Reply:
Exact equality will prevent a lot of maintenance issues.  Inexact comparison is merely tolerated.

Reply:
I agree that 1 percent doesn't look like a small tolerance, but it's actually enough for what we are testing here. If we decrease tolerance we will get unstable tests, which is much worse. 

I know that exact equality is always better choice than approximate equality. And I've spent a lot of time to do that, but couldn't get the desired result. And only after looking to source code of camera translation, I realized that it should be due to this [line](https://github.com/qt/qt3d/blob/82b59c89e282dcc04a01727e355d284c128752ab/src/render/frontend/qcamera.cpp#L607) and possibly we can't do much here or it will take a lot of time.

To prove my point here's the log data that I got:

Before translating camera towards X+ coordinate:
camera position: (0, 0, 10)
view center: (0, 0, 0)
view vector: (0, 0, -10)

After translate by (0.5, 0, 0):
camera position: (0.5, 0.000123389, 10)

Reply:
We need to find a balance between stability and sensitivity.  1 percent is so large that the stability may be false.  Graphical code uses single precision and has 1e-6 precision.  1 percent is 4 orders of magnitude larger than that.

I need more arguments to be convinced to use 1 percent for the tolerance. Developing the arguments will help us clarify what is a proper tolerance.

Reply:
We can decrease tolerance to 1e-3 then. What do you think?

Reply:
We move camera by the value of 0.5, so tolerance of 0.01 will always detect if camera did/didn't move. Of course 0.5 can be changed in the future to a bigger or smaller value, but I am pretty sure that we will never need to test if translation work on such small values as 0.01.

But yeah the smaller the better. We can decrease to 1e-3. And if we want to make it even more precise, it will take some time, that I think won't really benefit the project

Reply:
When saying something like "I am pretty sure ...", we usually mean that "from our experience, it must ..."  But based on what we have discussed so far, we probably do not have sound experience to serve as a background reference to say how a camera or Qt should behave.  But please let me know if you have a hidden argument with that sentence.

When we do not have sufficient experience to make decision, the best too is our reasoning faculty.  We need good (logical) arguments to exercise that.  I know the round-off of single precision floating point is 1e-6 and 1.e-2 and 1.e-4 are 2 to 4 orders of magnitude larger than that.  If the error in Qt is resulted from round-off, it should not be as big as 1.e-4.

I hope to learn something I didn't know from your arguments.  And based on the arguments we decide what to do next.  (I only need arguments.  I am fine for you to do more experiments, but that is not my point.)

Reply:
So I found the reason why that's happening. When camera's created for some reason it's upVector is not (0, 1, 0) but some strange value like (0.0001, 0.9999, 0). That's why when we translate camera, these strange values appear...

I prepared a commit solve this, I'm going to push it soon. This commit also will solve the issue when camera was reset, but it's rotation is strange

Reply:
That adds up.  If the input has an error of 1.e-4, the error of 1.23e-4 in the result makes sense.  We can wait for your change, but feel free to discuss more if you want.